File: iosApp/iosApp/Pages/Map/MapHttpInterceptor.swift
//
//  MapHttpInterceptor.swift
//  iosApp
//
//  Created by Brady, Kayla on 7/24/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Foundation
import MapboxMaps

class MapHttpInterceptor: HttpServiceInterceptorInterface {
    var updateLastErrorTimestamp: () -> Void

    init(updateLastErrorTimestamp: @escaping () -> Void) {
        self.updateLastErrorTimestamp = updateLastErrorTimestamp
    }

    func onRequest(for request: HttpRequest, continuation: @escaping HttpServiceInterceptorRequestContinuation) {
        // no-op
        continuation(HttpRequestOrResponse.fromHttpRequest(request))
    }

    func onResponse(for response: HttpResponse, continuation: @escaping HttpServiceInterceptorResponseContinuation) {
        switch response.result {
        case let .success(data):
            // 401 happens here
            if data.code == 401 {
                updateLastErrorTimestamp()
            }
        case let .failure(error):
            break
        }
        continuation(response)
    }
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/map/MapHttpInterceptor.kt
package com.mbta.tid.mbta_app.android.map

import com.mapbox.common.HttpRequest
import com.mapbox.common.HttpRequestOrResponse
import com.mapbox.common.HttpResponse
import com.mapbox.common.HttpServiceInterceptorInterface
import com.mapbox.common.HttpServiceInterceptorRequestContinuation
import com.mapbox.common.HttpServiceInterceptorResponseContinuation

class MapHttpInterceptor(val updateLastErrorTimestamp: () -> Unit) :
    HttpServiceInterceptorInterface {
    override fun onRequest(
        request: HttpRequest,
        continuation: HttpServiceInterceptorRequestContinuation
    ) {
        continuation.run(HttpRequestOrResponse(request))
    }

    override fun onResponse(
        response: HttpResponse,
        continuation: HttpServiceInterceptorResponseContinuation
    ) {

        val responseData = response.result.value

        if (responseData != null) {
            if (responseData.code == 401) {
                updateLastErrorTimestamp()
            }
        }
        continuation.run(response)
    }
}
